docs/design.md

```diff
@@ -0,0 +1,156 @@
+
+When features from Hypper can be upstreamed into Helm they will be. In that way,
+Hypper can act as a testing ground for Helm. We do not expect that every feature
+will fit in upstream Helm.
+
+A second element in the difference is that Helm, as a package manager, does not
+handle environment deployment operations. Those are left to higher level tools
+like Helmfile (a push model like Ansible) and the Flux Helm Operator (a pull
+model like Chef). Helm expects higher level tools will use Helm and charts for
+operating in different environments.
+
+Hypper will move will make some assumptions that move it more towards the
+environment management. This is in a different way from Helmfile or the Flux
+Helm Operator but further than Helm will go.
+
+## Directory Structure
+
+The directory structure is inspired by Helm with the following layout.
+
+```
+.
+├── cmd
+│   └── hypper
+├── docs
+└── pkg
+    └── action 
+    └── ... 
+```
+
+- The _cmd/hypper_ directory contains the client application. This CLI for mac,
+  linux, and windows is both a functional application and reference application
+  of the library.
+- _docs_ contains documentation on the design and usage of hypper.
+- The _pkg_ directory contains a library meant to be imported by _cmd/hypper_ and
+  other applications. The functionality needed is implemented here.
+- _pkg/actions_ contains the main functionality used by _cmd/hypper_.
+
+This is the same layout as Helm itself.
+
+## Design Intent
+
+The following outline the initial major intentions of hypper.
+
+### Modern Terminal UI
+
+Colors and icons are often present in modern terminals. Sometimes referred to
+as eye candy, these features can improve user experience by drawing attention to
+important things.
+
+### Optional Dependencies
+
+In Helm there are two ways to handle dependencies.
+
+1. Explicitly declare other charts as dependencies. This will cause those charts
+   to be installed. When the templates are rendered they happen in the same
+   sandboxed environment.
+2. Install charts separately and use values to connect them to each other. For
+   example, install PostreSQL with one chart and then pass connection information
+   into another chart using a Secret or values that generate a Secret.
+
+Hypper intends to add a 3rd form of dependency via an optional dependency. For
+example, you are developing something that requires a prometheus CRD. You can
+depend on a chart that provides that CRD but does not provide prometheus. The
+CRD chart or your chart can declare an optional dependency on a chart that
+installs Prometheus. When using hypper to install your chart you will be asked
+about installing the optional dependencies. In development you can choose not
+to install the dependencies while in production you can choose to install them.
+
+The optional dependency handling will extend to all commands.
+
+Optional dependencies will be specified using Helm chart annotations.
+
+### Shared Dependencies
```

Comment:
isnt this overlapping with point 2 on optional dependencies? Sounds similar to me!

Reply:
Optional dependencies are ones that may not be installed. Shared dependencies are those that will be installed but they will be shared so we don't have multiple versions of something (e.g., istio) installed.

Reply:
For naming, I suggest to also get inspired by package managers (for example python's pip), this would be clearer:
- Shared depencies -> System packages (to be opposed with "user" packages, which would here be better named as "namespaced packages", IMO)
-  Optional -> Recommends/Requires (This is more explicit).
WDYT?

Reply:
I like the idea of system packages and namespaced packages.

> Optional -> Recommends/Requires (This is more explicit).

Two thoughts come to mind...

1. We aren't building a new package manager. We are building on top of Helm and charts. So, we need to work around and with what we have today. The current `dependencies` list is equivalent to required ones.
2. The language around optional/recommends is something we can work out when we get to this feature. We should look at what others are doing in both the system and programming language package managers that are popular. From there we can make naming decisions around what will be easier for people to understand and figure out.